tsdb/tsm1/engine_schema.go

```diff
@@ -230,15 +230,26 @@ func (e *Engine) tagValuesPredicate(ctx context.Context, orgID, bucketID influxd
 			continue
 		}
 
-		keybuf = models.AppendMakeKey(keybuf[:0], orgBucketEsc, tags)
+		// orgBucketEsc is already escaped, so no need to use models.AppendMakeKey, which
+		// unescapes and escapes the value again. The degenerate case is if the orgBucketEsc
+		// has escaped values, causing two allocations per key
+		keybuf = append(keybuf[:0], orgBucketEsc...)
+		keybuf = tags.AppendHashKey(keybuf)
 		sfkey = AppendSeriesFieldKeyBytes(sfkey[:0], keybuf, tags.Get(models.FieldKeyTagKeyBytes))
 
-		values := e.Cache.Values(sfkey)
-		stats.ScannedValues += values.Len()
-		stats.ScannedBytes += values.Len() * 8 // sizeof timestamp
+		found := false
+		_ = e.Cache.ApplyEntryFnForKey(sfkey, func(entry *entry) error {
+			stats.ScannedValues += entry.values.Len()
+			stats.ScannedBytes += entry.values.Len() * 8 // sizeof timestamp
 
-		if values.Contains(start, end) {
-			tsmValues[string(curVal)] = struct{}{}
+			if entry.values.Contains(start, end) {
```

Comment:
I'm not sure I like this. Seems fine, but also seems a little off. The commit is adding new APIs on `entry` to avoid dipping into `values` unsafely, and then here you dip into `values`, albeit safely because you have a read lock via `entry.apply()`.

I think there is a little muddling going on with exported/unexported and locking on `entry`. Probably historical cruft....

Either way, I wonder if either `entry.apply` shouldn't take a lock (and leave the locking to the function argument), or perhaps if there should be an `entry.contains` that doesn't do any locking on `entry` and dips into `entry.values.Contains` for you.

Just some thoughts anyway..

Reply:
In the end I didn't like submitting this change either. I'll remove the API and use the same `MaxTime` and `Contains` APIs

Reply:
(with the proposed changes to do the `O(N)` operations.